dex/pair/src/bot_protection.rs

```diff
@@ -0,0 +1,196 @@
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use crate::config;
+use crate::contexts::add_liquidity::AddLiquidityContext;
+use crate::contexts::base::*;
+use crate::contexts::remove_liquidity::RemoveLiquidityContext;
+use crate::contexts::swap::SwapContext;
+
+const PERCENT_MAX: u64 = 100_000;
+
+#[derive(TopEncode, TopDecode, TypeAbi)]
+pub struct BPConfig {
+    protect_stop_block: u64,
+    volume_percent: u64,
+    max_num_actions_per_address: u64,
+}
+
+#[elrond_wasm::module]
+pub trait BPModule: config::ConfigModule + token_send::TokenSendModule {
+    fn require_can_proceed_swap(&self, ctx: &SwapContext<Self::Api>) {
+        if self.bp_swap_config().is_empty() {
+            return;
+        }
+
+        let bp_config = self.bp_swap_config().get();
+        let current_block = self.blockchain().get_block_nonce();
+        if bp_config.protect_stop_block < current_block {
+            return;
+        }
+
+        let reserve_in = ctx.get_reserve_in();
+        let reserve_out = ctx.get_reserve_out();
+        if reserve_in == &0u64 && reserve_out == &0u64 {
+            return;
+        }
+
+        let caller = ctx.get_caller();
+        let num_swaps = self.num_swaps_by_address(caller).get();
+        require!(
+            num_swaps < bp_config.max_num_actions_per_address,
+            "too many swaps by address"
+        );
+
+        let final_in = ctx.get_final_input_amount();
+        let amount_in_percent = final_in * PERCENT_MAX / reserve_in;
+        require!(
+            amount_in_percent < bp_config.volume_percent,
+            "swap amount in too large"
+        );
+
+        let final_out = ctx.get_final_output_amount();
+        let amount_out_percent = final_out * PERCENT_MAX / reserve_out;
+        require!(
+            amount_out_percent < bp_config.volume_percent,
+            "swap amount out too large"
+        );
+
+        self.num_swaps_by_address(caller).set(&(num_swaps + 1));
+    }
+
+    fn require_can_proceed_remove(&self, ctx: &RemoveLiquidityContext<Self::Api>) {
```

Comment:
why do we need protection for remove and add ?

Reply:
because swap works with volume percents, one whale can manipulate the volumes and make high volume swaps by after adding much volume. this protection takes care of that by not allowing anyone to high add or remove (thus manipulating allowed the volume amount on swap - being fixed percent of total)